File: web/transports/audio.mjs
// Audio-modem transport — send SPORE envelopes *as sound*. Two laptops on a
// table, a radio's speaker into another radio's mic, a room full of phones: the
// medium is broadcast-only and has no address, so everyone in earshot hears every
// frame and the envelope's own `dest` decides who cares.
//
// This is a faithful port of the Rust bridge (`src/bridge/audio.rs`): a 16-tone
// FSK modem with a Goertzel receiver, **bit-compatible** with it — a browser tab
// and a native `spore-audio` pipe exchange real envelopes over the air.
//
//   Wire (per frame):  SYNC(6 sym) · LEN(2 B) · PAYLOAD · CRC(4 B)
//   16 tones ⇒ 4 bits/symbol, high nibble first; CRC = SHA-256(LEN‖PAYLOAD)[0..4].
//
//   const t = await AudioModemTransport.open(); // asks for the mic
//   hub.addTransport(t);
import { Transport } from '../spore.mjs';

export const SAMPLE_RATE = 48000; // sound cards universally support 48 kHz
export const SYMBOL_LEN = 1024; // bin-aligned: SR/SYMBOL_LEN = 46.875 Hz
const TONES = 16; // 4 bits per symbol
const BASE_BIN = 32; // 1500 Hz
const SPACING_BIN = 4; // 187.5 Hz between tones
const SYNC = [15, 0, 15, 0, 12, 3];
const AMPLITUDE = 0.5;
export const MAX_FRAME_PAYLOAD = 4096;

const toneFreq = (s) => ((BASE_BIN + s * SPACING_BIN) * SAMPLE_RATE) / SYMBOL_LEN;

// ---- minimal synchronous SHA-256 (for the 4-byte frame CRC only) -------------
// The envelope itself is Ed25519-signed; this CRC only rejects a mis-heard frame,
// and must byte-match Rust's `Sha256::digest(frame)[..4]`.
const K = new Uint32Array([
  0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
  0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
  0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
  0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
  0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
  0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
  0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
  0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
]);
function sha256(bytes) {
  const h = new Uint32Array([
    0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19,
  ]);
  const ml = bytes.length * 8;
  const withPad = new Uint8Array((((bytes.length + 8) >> 6) + 1) << 6);
  withPad.set(bytes);
  withPad[bytes.length] = 0x80;
  const dv = new DataView(withPad.buffer);
  dv.setUint32(withPad.length - 4, ml >>> 0);
  dv.setUint32(withPad.length - 8, Math.floor(ml / 0x100000000));
  const w = new Uint32Array(64);
  const rotr = (x, n) => (x >>> n) | (x << (32 - n));
  for (let i = 0; i < withPad.length; i += 64) {
    for (let t = 0; t < 16; t++) w[t] = dv.getUint32(i + t * 4);
    for (let t = 16; t < 64; t++) {
      const s0 = rotr(w[t - 15], 7) ^ rotr(w[t - 15], 18) ^ (w[t - 15] >>> 3);
      const s1 = rotr(w[t - 2], 17) ^ rotr(w[t - 2], 19) ^ (w[t - 2] >>> 10);
      w[t] = (w[t - 16] + s0 + w[t - 7] + s1) | 0;
    }
    let [a, b, c, d, e, f, g, hh] = h;
    for (let t = 0; t < 64; t++) {
      const S1 = rotr(e, 6) ^ rotr(e, 11) ^ rotr(e, 25);
      const ch = (e & f) ^ (~e & g);
      const t1 = (hh + S1 + ch + K[t] + w[t]) | 0;
      const S0 = rotr(a, 2) ^ rotr(a, 13) ^ rotr(a, 22);
      const maj = (a & b) ^ (a & c) ^ (b & c);
      const t2 = (S0 + maj) | 0;
      hh = g; g = f; f = e; e = (d + t1) | 0; d = c; c = b; b = a; a = (t1 + t2) | 0;
    }
    h[0] = (h[0] + a) | 0; h[1] = (h[1] + b) | 0; h[2] = (h[2] + c) | 0; h[3] = (h[3] + d) | 0;
    h[4] = (h[4] + e) | 0; h[5] = (h[5] + f) | 0; h[6] = (h[6] + g) | 0; h[7] = (h[7] + hh) | 0;
  }
  const out = new Uint8Array(32);
  new DataView(out.buffer).setUint32(0, h[0] >>> 0);
  for (let i = 0; i < 8; i++) new DataView(out.buffer).setUint32(i * 4, h[i] >>> 0);
  return out;
}

// ---- modulation (bytes -> PCM) ----------------------------------------------
export function modulate(payload) {
  const len = Math.min(payload.length, 0xffff);
  const frame = new Uint8Array(2 + len);
  frame[0] = len >> 8;
  frame[1] = len & 0xff;
  frame.set(payload.subarray(0, len), 2);
  const crc = sha256(frame).subarray(0, 4);
  const framed = new Uint8Array(frame.length + 4);
  framed.set(frame);
  framed.set(crc, frame.length);

  const symbols = SYNC.slice();
  for (const b of framed) symbols.push(b >> 4, b & 0x0f);

  const out = new Float32Array(symbols.length * SYMBOL_LEN);
  let o = 0;
  for (const s of symbols) {
    const f = toneFreq(s);
    for (let n = 0; n < SYMBOL_LEN; n++) {
      out[o++] = AMPLITUDE * Math.sin((2 * Math.PI * f * n) / SAMPLE_RATE);
    }
  }
  return out;
}

// ---- demodulation (PCM -> bytes) --------------------------------------------
function goertzel(buf, off, freq) {
  const k = Math.round((SYMBOL_LEN * freq) / SAMPLE_RATE);
  const w = (2 * Math.PI * k) / SYMBOL_LEN;
  const coeff = 2 * Math.cos(w);
  let s1 = 0, s2 = 0;
  for (let i = 0; i < SYMBOL_LEN; i++) {
    const s0 = buf[off + i] + coeff * s1 - s2;
    s2 = s1;
    s1 = s0;
  }
  return Math.sqrt(Math.max(s1 * s1 + s2 * s2 - coeff * s1 * s2, 0));
}
function symbolAt(buf, off) {
  if (off + SYMBOL_LEN > buf.length) return null;
  let best = 0, bestMag = 0, sum = 1e-9;
  for (let s = 0; s < TONES; s++) {
    const m = goertzel(buf, off, toneFreq(s));
    sum += m;
    if (m > bestMag) { bestMag = m; best = s; }
  }
  return [best, bestMag / sum];
}
function syncMatches(buf, off) {
  for (let i = 0; i < SYNC.length; i++) {
    const r = symbolAt(buf, off + i * SYMBOL_LEN);
    if (!r || r[0] !== SYNC[i] || r[1] <= 0.3) return false;
  }
  return true;
}
function readBytes(buf, start, count) {
  const out = new Uint8Array(count);
  for (let j = 0; j < count; j++) {
    const hi = symbolAt(buf, start + 2 * j * SYMBOL_LEN);
    const lo = symbolAt(buf, start + (2 * j + 1) * SYMBOL_LEN);
    if (!hi || !lo) return null;
    out[j] = (hi[0] << 4) | lo[0];
  }
  return out;
}
function decodeFrame(buf, off) {
  const dataOff = off + SYNC.length * SYMBOL_LEN;
  const lenB = readBytes(buf, dataOff, 2);
  if (!lenB) return null;
  const payloadLen = (lenB[0] << 8) | lenB[1];
  if (payloadLen > MAX_FRAME_PAYLOAD) return null;
  const frameBytes = 2 + payloadLen + 4;
  const framed = readBytes(buf, dataOff, frameBytes);
  if (!framed) return null;
  const body = framed.subarray(0, frameBytes - 4);
  const crc = sha256(body).subarray(0, 4);
  for (let i = 0; i < 4; i++) if (crc[i] !== framed[frameBytes - 4 + i]) return null;
  const end = dataOff + 2 * frameBytes * SYMBOL_LEN;
  return [body.subarray(2), end]; // strip the 2 LEN bytes
}

/** Streaming receiver: push PCM as it arrives, get back completed payloads. */
export class Demod {
  constructor() {
    this.buf = new Float32Array(0);
  }
  push(samples) {
    const merged = new Float32Array(this.buf.length + samples.length);
    merged.set(this.buf);
    merged.set(samples, this.buf.length);
    this.buf = merged;
    const out = [];
    const step = SYMBOL_LEN >> 3;
    let off = 0, consumed = 0;
    while (off + SYNC.length * SYMBOL_LEN <= this.buf.length) {
      if (syncMatches(this.buf, off)) {
        const r = decodeFrame(this.buf, off);
        if (r) { out.push(r[0]); off = r[1]; consumed = r[1]; continue; }
      }
      off += step;
    }
    const maxFrame = (SYNC.length + 2 * (2 + MAX_FRAME_PAYLOAD + 4)) * SYMBOL_LEN;
    const keepFrom = Math.max(consumed, this.buf.length - maxFrame);
    if (keepFrom > 0) this.buf = this.buf.slice(keepFrom);
    return out;
  }
}

// ---- browser transport (mic + speaker) --------------------------------------
export class AudioModemTransport extends Transport {
  constructor(ctx, stream) {
    super();
    this.ctx = ctx;
    this.stream = stream;
    this.demod = new Demod();

    const src = ctx.createMediaStreamSource(stream);
    // ScriptProcessor is deprecated but works everywhere including file://; the
    // 0-gain sink lets it fire without echoing the mic to the speakers.
    const proc = ctx.createScriptProcessor(4096, 1, 1);
    this.proc = proc;
    proc.onaudioprocess = (ev) => {
      const inp = ev.inputBuffer.getChannelData(0);
      for (const frame of this.demod.push(new Float32Array(inp))) this.receive(frame);
    };
    const mute = ctx.createGain();
    mute.gain.value = 0;
    src.connect(proc);
    proc.connect(mute);
    mute.connect(ctx.destination);
  }

  /** Open the mic (asks permission) at 48 kHz and start listening. */
  static async open() {
    if (!navigator.mediaDevices?.getUserMedia) throw new Error('microphone not available');
    const AC = window.AudioContext || window.webkitAudioContext;
    const ctx = new AC({ sampleRate: SAMPLE_RATE });
    if (ctx.sampleRate !== SAMPLE_RATE) {
      // A few platforms ignore the hint; the modem needs exactly 48 kHz.
      await ctx.close();
      throw new Error(`audio device is ${ctx.sampleRate} Hz; the modem needs ${SAMPLE_RATE} Hz`);
    }
    const stream = await navigator.mediaDevices.getUserMedia({
      audio: { echoCancellation: false, noiseSuppression: false, autoGainControl: false },
    });
    await ctx.resume();
    return new AudioModemTransport(ctx, stream);
  }

  send(bytes) {
    const pcm = modulate(bytes);
    const audioBuf = this.ctx.createBuffer(1, pcm.length, SAMPLE_RATE);
    audioBuf.getChannelData(0).set(pcm);
    const node = this.ctx.createBufferSource();
    node.buffer = audioBuf;
    node.connect(this.ctx.destination);
    node.start();
  }

  close() {
    try { this.proc.disconnect(); } catch { /* */ }
    try { this.stream.getTracks().forEach((t) => t.stop()); } catch { /* */ }
    try { this.ctx.close(); } catch { /* */ }
  }
}

File: src/bridge/audio.rs
//! Audio modem bridge (ggwave-style) — send SPORE envelopes *as sound*.
//!
//! Data-over-sound: two phones on a table, a radio's speaker to another radio's
//! mic, a room full of laptops. The medium is broadcast-only and has no address
//! (`U = ()`), so everyone in earshot hears every frame and the envelope's own
//! `dest` decides who cares — exactly the "null address" row of the bridge
//! matrix.
//!
//! ## What's portable vs. platform
//! Everything that turns bytes into samples and back lives here and is fully
//! tested: a 16-tone FSK modem ([`modulate`]) with a Goertzel receiver
//! ([`demodulate`] / [`Demod`]). The only platform-specific part is moving `f32`
//! PCM to and from a speaker/mic. [`run_pipe`] does that over stdin/stdout so you
//! can wire any sound backend without a Rust audio dependency, e.g.:
//!
//! ```text
//! # receive: mic -> sox -> spore ; transmit: spore -> sox -> speaker
//! sox -d -t f32 -r 48000 -c 1 - | spore-audio | sox -t f32 -r 48000 -c 1 - -d
//! ```
//!
//! In the browser the same [`modulate`]/[`demodulate`] compiled to wasm feed a
//! `WebAudio` `ScriptProcessor`/`AudioWorklet`.
//!
//! ## Wire format (per frame)
//! `SYNC(6 symbols)` · `LEN(2 bytes)` · `PAYLOAD(LEN bytes)` · `CRC(4 bytes)`.
//! Bytes are sent high-nibble-first, one 4-bit symbol per tone. The CRC is the
//! shared `SHA-256(frame)[0..4]` tail (`bridge::crc_append`).

use crate::bridge::{crc_append, crc_check};
use core::f32::consts::PI;

/// Samples per second the modem assumes. Sound cards universally support 48 kHz.
pub const SAMPLE_RATE: u32 = 48_000;
/// Samples per symbol. Bin-aligned to `SAMPLE_RATE / SYMBOL_LEN = 46.875 Hz` so
/// Goertzel lands exactly on each tone. ~21 ms/symbol → ~47 symbols/s.
pub const SYMBOL_LEN: usize = 1024;
/// 16 tones ⇒ 4 bits per symbol ⇒ ~23 bytes/s.
const TONES: usize = 16;

/// What this link will carry of *other people's file chunks*, per second — see
/// [`crate::bridge::hub::Hub::register_limited`].
///
/// Zero. At ~23 bytes/s one 1336-byte chunk occupies the channel for a minute,
/// so a sound link that served bulk would serve nothing else. It still carries
/// messages, announces and manifests at full speed, which is what an audio link
/// is actually for: telling the mesh you exist and what you have. Whoever wants
/// the chunks asks again and a faster path answers.
pub const BULK_BYTES_PER_SEC: u32 = 0;
const BASE_BIN: usize = 32; // 1500 Hz
const SPACING_BIN: usize = 4; // 187.5 Hz between tones — well separated at this window
const SYNC: [u8; 6] = [15, 0, 15, 0, 12, 3];
const AMPLITUDE: f32 = 0.5;

fn tone_freq(symbol: u8) -> f32 {
    ((BASE_BIN + symbol as usize * SPACING_BIN) * SAMPLE_RATE as usize) as f32 / SYMBOL_LEN as f32
}

/// Modulate a payload into mono `f32` PCM at [`SAMPLE_RATE`]. Prepend/append
/// silence yourself if your output needs a lead-in.
pub fn modulate(payload: &[u8]) -> Vec<f32> {
    // frame = LEN(2) || payload, then a 4-byte CRC tail.
    let len = payload.len().min(0xFFFF);
    let mut frame = vec![(len >> 8) as u8, (len & 0xFF) as u8];
    frame.extend_from_slice(&payload[..len]);
    let frame = crc_append(&frame);

    let mut symbols: Vec<u8> = SYNC.to_vec();
    for &b in &frame {
        symbols.push(b >> 4);
        symbols.push(b & 0x0F);
    }

    let mut out = Vec::with_capacity(symbols.len() * SYMBOL_LEN);
    for &s in &symbols {
        let f = tone_freq(s);
        for n in 0..SYMBOL_LEN {
            let t = n as f32 / SAMPLE_RATE as f32;
            out.push(AMPLITUDE * (2.0 * PI * f * t).sin());
        }
    }
    out
}

// Goertzel magnitude at a bin-aligned frequency over exactly one symbol window.
fn goertzel(samples: &[f32], freq: f32) -> f32 {
    let n = samples.len() as f32;
    let k = (n * freq / SAMPLE_RATE as f32).round();
    let w = 2.0 * PI * k / n;
    let coeff = 2.0 * w.cos();
    let (mut s1, mut s2) = (0.0f32, 0.0f32);
    for &x in samples {
        let s0 = x + coeff * s1 - s2;
        s2 = s1;
        s1 = s0;
    }
    (s1 * s1 + s2 * s2 - coeff * s1 * s2).max(0.0).sqrt()
}

// Best-matching tone for one symbol window, with its confidence (peak / total).
fn symbol_at(samples: &[f32], off: usize) -> Option<(u8, f32)> {
    if off + SYMBOL_LEN > samples.len() {
        return None;
    }
    let win = &samples[off..off + SYMBOL_LEN];
    let (mut best, mut best_mag, mut sum) = (0u8, 0.0f32, 1e-9f32);
    for s in 0..TONES as u8 {
        let m = goertzel(win, tone_freq(s));
        sum += m;
        if m > best_mag {
            best_mag = m;
            best = s;
        }
    }
    Some((best, best_mag / sum))
}

fn sync_matches(samples: &[f32], off: usize) -> bool {
    for (i, &want) in SYNC.iter().enumerate() {
        match symbol_at(samples, off + i * SYMBOL_LEN) {
            Some((sym, conf)) if sym == want && conf > 0.30 => {}
            _ => return false,
        }
    }
    true
}

fn read_bytes(samples: &[f32], start_off: usize, count: usize) -> Option<Vec<u8>> {
    let mut bytes = Vec::with_capacity(count);
    for j in 0..count {
        let (hi, _) = symbol_at(samples, start_off + (2 * j) * SYMBOL_LEN)?;
        let (lo, _) = symbol_at(samples, start_off + (2 * j + 1) * SYMBOL_LEN)?;
        bytes.push((hi << 4) | lo);
    }
    Some(bytes)
}

/// The largest payload a single frame carries (keeps a bogus LEN from allocating
/// wildly). Bigger objects fragment above the bridge, like every other medium.
pub const MAX_FRAME_PAYLOAD: usize = 4096;

// Try to decode one frame whose SYNC starts at `off`. Returns the payload and
// the sample offset just past the frame, or `None` if it wasn't a real frame.
fn decode_frame(samples: &[f32], off: usize) -> Option<(Vec<u8>, usize)> {
    let data_off = off + SYNC.len() * SYMBOL_LEN;
    let len_bytes = read_bytes(samples, data_off, 2)?;
    let payload_len = ((len_bytes[0] as usize) << 8) | len_bytes[1] as usize;
    if payload_len > MAX_FRAME_PAYLOAD {
        return None;
    }
    let frame_bytes = 2 + payload_len + 4; // LEN + payload + CRC tail
    let frame = read_bytes(samples, data_off, frame_bytes)?;
    let body = crc_check(&frame)?; // rejects false syncs and corruption
    let end = data_off + (2 * frame_bytes) * SYMBOL_LEN;
    Some((body[2..].to_vec(), end)) // strip the 2 LEN bytes
}

/// Demodulate a whole PCM buffer, returning every payload recovered from it.
/// Scans for the sync word at a fraction of a symbol so it tolerates arbitrary
/// lead-in silence and coarse timing offset.
pub fn demodulate(samples: &[f32]) -> Vec<Vec<u8>> {
    let step = SYMBOL_LEN / 8;
    let mut out = Vec::new();
    let mut off = 0usize;
    while off + SYNC.len() * SYMBOL_LEN <= samples.len() {
        if sync_matches(samples, off) {
            if let Some((payload, end)) = decode_frame(samples, off) {
                out.push(payload);
                off = end; // jump past the frame we just read
                continue;
            }
        }
        off += step;
    }
    out
}

/// Streaming receiver: push PCM as it arrives, get back completed payloads.
/// Buffers just enough to span one in-flight frame.
#[derive(Default)]
pub struct Demod {
    buf: Vec<f32>,
    /// Offsets below this have been tested for sync and cannot match later.
    ///
    /// Without it `push` restarted at zero every call, so the work per call was
    /// proportional to the whole retained buffer rather than to the new samples.
    /// Measured before the fix: a 100 ms push cost 13 ms with 1 s buffered, 94 ms
    /// with 6 s, and the buffer caps at 175 s — about 27x real time, one core
    /// saturated for as long as the room is noisy. Any sound that does not decode
    /// gets you there, with no key and no protocol participation (S-031).
    scanned: usize,
}

/// Samples one maximal frame can span — the most the buffer ever needs to hold.
fn max_frame_samples() -> usize {
    (SYNC.len() + 2 * (2 + MAX_FRAME_PAYLOAD + 4)) * SYMBOL_LEN
}

impl Demod {
    pub fn new() -> Self {
        Demod { buf: Vec::new(), scanned: 0 }
    }

    /// Append newly captured samples; returns any frames that completed.
    pub fn push(&mut self, samples: &[f32]) -> Vec<Vec<u8>> {
        self.buf.extend_from_slice(samples);
        let mut out = Vec::new();
        let step = SYMBOL_LEN / 8;
        // Resume where the last call stopped. An offset that had a full sync
        // window and did not match cannot start matching once *more* samples
        // arrive after it, so retesting it is pure waste.
        let mut off = self.scanned;
        let mut consumed = 0usize;
        while off + SYNC.len() * SYMBOL_LEN <= self.buf.len() {
            if sync_matches(&self.buf, off) {
                if let Some((payload, end)) = decode_frame(&self.buf, off) {
                    out.push(payload);
                    off = end;
                    consumed = end;
                    continue;
                }
                // Sync is here but the frame did not decode. Usually that means
                // the body is still arriving, so hold this offset and retry on the
                // next push — advancing past it would drop every frame whose sync
                // lands near the end of a chunk, which is most of them. Only give
                // up once the buffer is long enough that even a maximal frame
                // would have fitted, which bounds the stall.
                if self.buf.len() - off < max_frame_samples() {
                    break;
                }
            }
            off += step;
        }
        self.scanned = off;
        // Drop everything we've fully consumed; keep a tail that might hold the
        // start of a frame still being received. Cap the buffer so a silent mic
        // can't grow it without bound.
        let keep_from = consumed.max(self.buf.len().saturating_sub(max_frame_samples()));
        if keep_from > 0 {
            self.buf.drain(..keep_from);
            // Offsets shifted down by exactly what was removed.
            self.scanned = self.scanned.saturating_sub(keep_from);
        }
        out
    }
}

/// Sound-card-agnostic runner: read mono `f32` PCM from stdin (the mic), demod
/// and feed frames to the shared node; write modulated `f32` PCM of outbound
/// forwards to stdout (the speaker). Pipe both ends through `sox`/`ffmpeg`/
/// `pw-cat` — that pipe is the only platform-specific glue.
#[cfg(not(target_arch = "wasm32"))]
pub fn run_pipe(
    hub: crate::bridge::hub::Shared,
    iface: crate::Iface,
    rx: std::sync::mpsc::Receiver<crate::Forward>,
) -> std::io::Result<()> {
    use std::io::{Read, Write};
    use std::sync::mpsc::TryRecvError;

    println!("  [audio] iface {iface} — f32 PCM {SAMPLE_RATE} Hz mono on stdin/stdout");
    let mut stdin = std::io::stdin().lock();
    let mut stdout = std::io::stdout().lock();
    let mut demod = Demod::new();
    let mut raw = [0u8; 4096];

    loop {
        // Transmit: modulate any pending forwards to the speaker.
        loop {
            match rx.try_recv() {
                Ok(f) => {
                    let bytes = match f {
                        crate::Forward::Flood { bytes, .. } => bytes,
                        crate::Forward::Directed { bytes, .. } => bytes,
                    };
                    let pcm = modulate(&bytes);
                    let mut out = Vec::with_capacity(pcm.len() * 4);
                    for s in pcm {
                        out.extend_from_slice(&s.to_le_bytes());
                    }
                    stdout.write_all(&out)?;
                    stdout.flush()?;
                }
                Err(TryRecvError::Empty) => break,
                Err(TryRecvError::Disconnected) => return Ok(()),
            }
        }
        // Receive: pull a chunk of mic PCM and demod it.
        let n = stdin.read(&mut raw)?;
        if n == 0 {
            return Ok(()); // mic stream ended
        }
        let mut samples = Vec::with_capacity(n / 4);
        for c in raw[..n / 4 * 4].chunks_exact(4) {
            samples.push(f32::from_le_bytes([c[0], c[1], c[2], c[3]]));
        }
        for frame in demod.push(&samples) {
            hub.on_rx(iface, &frame, None);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // A cheap deterministic PRNG so the test is reproducible without deps.
    fn noise(seed: &mut u64) -> f32 {
        *seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        ((*seed >> 40) as f32 / (1u64 << 24) as f32) - 0.5 // ~[-0.5, 0.5)
    }

    #[test]
    fn roundtrip_clean() {
        let msg = b"SPORE over sound";
        let pcm = modulate(msg);
        let got = demodulate(&pcm);
        assert_eq!(got, vec![msg.to_vec()]);
    }

    #[test]
    fn roundtrip_with_silence_and_noise() {
        let msg = b"the dam holds; meet at the north pier";
        let mut pcm = vec![0.0f32; 777]; // arbitrary (non-symbol-aligned) lead-in
        pcm.extend(modulate(msg));
        // `repeat_n` would be tidier but is 1.82; MSRV is 1.75.
        pcm.extend(std::iter::repeat_n(0.0, 500));
        // Add low-level channel noise.
        let mut seed = 0xDEADBEEFu64;
        for s in pcm.iter_mut() {
            *s += 0.06 * noise(&mut seed);
        }
        let got = demodulate(&pcm);
        assert_eq!(got, vec![msg.to_vec()]);
    }

    #[test]
    fn two_frames_back_to_back() {
        let a = b"first";
        let b = b"second frame";
        let mut pcm = modulate(a);
        pcm.extend(modulate(b));
        let got = demodulate(&pcm);
        assert_eq!(got, vec![a.to_vec(), b.to_vec()]);
    }

    #[test]
    fn streaming_demod_in_chunks() {
        let msg = b"streamed in little pieces";
        let pcm = modulate(msg);
        let mut d = Demod::new();
        let mut got = Vec::new();
        for chunk in pcm.chunks(333) {
            got.extend(d.push(chunk));
        }
        assert_eq!(got, vec![msg.to_vec()]);
    }
}

#[cfg(test)]
mod scan_tests {
    use super::*;

    /// S-031. The cursor must not cost us frames: a payload split across many
    /// small pushes — which is what a real mic delivers — must still decode.
    #[test]
    fn a_frame_split_across_many_pushes_still_decodes() {
        for chunk in [64usize, 257, 1024, 4096] {
            let payload = b"the dam holds, and the pier is clear".to_vec();
            let mut pcm = vec![0.0f32; 3000]; // lead-in silence, as a real mic has
            pcm.extend(modulate(&payload));
            pcm.extend(vec![0.0f32; 3000]);
            let mut d = Demod::new();
            let mut got = Vec::new();
            for part in pcm.chunks(chunk) {
                got.extend(d.push(part));
            }
            assert_eq!(got, vec![payload.clone()], "lost the frame at chunk size {chunk}");
        }
    }

    /// Two frames back to back, so the cursor has to survive a consume + drain.
    #[test]
    fn two_frames_in_a_stream_both_decode() {
        let mut pcm = vec![0.0f32; 2000];
        pcm.extend(modulate(b"first"));
        pcm.extend(vec![0.0f32; 2000]);
        pcm.extend(modulate(b"second"));
        pcm.extend(vec![0.0f32; 2000]);
        let mut d = Demod::new();
        let mut got = Vec::new();
        for part in pcm.chunks(512) {
            got.extend(d.push(part));
        }
        assert_eq!(got, vec![b"first".to_vec(), b"second".to_vec()]);
    }

    /// The property the cursor exists for: work per push must not grow with the
    /// buffer. Noise never syncs, so the buffer fills to its cap; if the scan
    /// still restarted at zero this would grow without bound.
    #[test]
    fn scanning_does_not_redo_work_as_the_buffer_fills() {
        let chunk = SAMPLE_RATE as usize / 10;
        let mut x = 12345u32;
        let noise: Vec<f32> = (0..chunk)
            .map(|_| {
                x = x.wrapping_mul(1103515245).wrapping_add(12345);
                (x >> 16) as f32 / 32768.0 - 1.0
            })
            .collect();
        let mut d = Demod::new();
        for _ in 0..5 {
            d.push(&noise);
        }
        let early = std::time::Instant::now();
        d.push(&noise);
        let early = early.elapsed();
        for _ in 0..40 {
            d.push(&noise);
        }
        let late = std::time::Instant::now();
        d.push(&noise);
        let late = late.elapsed();
        // Before the cursor this ratio tracked the buffer length and reached ~30x.
        // Generous bound: timing on shared CI is noisy, and the point is that it
        // is flat rather than linear.
        assert!(
            late < early * 8 + std::time::Duration::from_millis(5),
            "push cost grew with the buffer: {early:?} early vs {late:?} late"
        );
    }
}
